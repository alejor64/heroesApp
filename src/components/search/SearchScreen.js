import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import HeroeCard from '../heroes/HeroeCard'
import { getHeroesByName } from '../../selector/getHeroesByName'

const SearchScreen = ({history}) => {

    const location = useLocation()
    const {q = ''} = queryString.parse(location.search)

    const [values, handleInputChange] = useForm({
        searchText: q
    })
    
    const {searchText} = values

    const heroesFilter = useMemo(() => getHeroesByName(q), [q])
    
    const handleSearch = (e) => {
        e.preventDefault()
        history.push(`?q=${searchText}`)
    }

    return (
        <div>
            <h1>Search</h1>
            <hr/>

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input 
                            type="text"
                            placeholder="Find your heroe"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    {
                        (q ==='') && 
                            <div className="alert alert-info">
                                Search a Heroe
                            </div>
                    }

                    {
                        (q !=='' && heroesFilter.length === 0) && 
                            <div className="alert alert-danger">
                                There is no a heroe with "{q}"
                            </div>
                    }

                    {
                        heroesFilter.map(heroe => (
                            <HeroeCard
                                key={heroe.id}
                                {...heroe}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}

export default SearchScreen
