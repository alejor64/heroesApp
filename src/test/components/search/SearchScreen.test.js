import React from 'react'
import '@testing-library/jest-dom'
import { mount } from 'enzyme'
import SearchScreen from '../../../components/search/SearchScreen'
import { MemoryRouter, Route } from 'react-router-dom'

describe('Pruebas en el componente <SearchScreen/>', () => {
    
    test('Debe mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-info').exists()).toBe(true)
    })
    
    test('Debe de mostrar a Batman y el input con el queryString', () => {
        const queryString = 'batman'
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${queryString}`]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('input').prop('value')).toBe(queryString)
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('HeroeCard').exists()).toBe(true)
    })
    
    test('Debe mostrar un error sino se encuentra el heroe', () => {
        const queryString = 'batman123'
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${queryString}`]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        )
        
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-danger').exists()).toBe(true)
        expect(wrapper.find('.alert-danger').text()).toBe(`There is no a heroe with "${queryString}"`)
    })
    
    test('Debe llamar el PUSH del History', () => {
        const history = {
            push: jest.fn()
        }
        const queryString = 'batman'
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${queryString}`]}>
                <Route path="/search" component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', {
            target: {
                name: "searchText",
                value: queryString
            }
        })

        wrapper.find('form').prop('onSubmit')({preventDefault(){}})
        expect(history.push).toHaveBeenCalledWith(`?q=${queryString}`)
    })
    
})
