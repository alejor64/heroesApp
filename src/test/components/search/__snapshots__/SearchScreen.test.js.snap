// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <SearchScreen/> Debe de mostrar a Batman y el input con el queryString 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                More about...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente <SearchScreen/> Debe mostrar un error sino se encuentra el heroe 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value="batman123"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a heroe with "
        batman123
        "
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente <SearchScreen/> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Heroe
      </div>
    </div>
  </div>
</div>
`;
